File: 2021/10/17/LongPressedName.java
/**
 * https://leetcode.com/problems/long-pressed-name/
 * @author Andrew Jarombek
 * @since 10/17/2021
 */

public class LongPressedName {
    public boolean isLongPressedName(String name, String typed) {
        if (name.equals(typed)) {
            return true;
        }

        if (name.length() > typed.length()) {
            return false;
        }

        int index = 0;
        Character prev = null;

        for (char c: name.toCharArray()) {
            if (prev == null || prev != c) {
                while (prev != null && index < typed.length() && typed.charAt(index) == prev) {
                    index++;
                }

                prev = c;
            }

            if (index == typed.length() || typed.charAt(index) != c) {
                return false;
            }

            index++;
        }

        while (prev != null && index < typed.length() && typed.charAt(index) == prev) {
            index++;
        }

        if (index < typed.length()) {
            return false;
        }

        return true;
    }
}

File: 2022/05/23/longPressedName.py
"""
https://leetcode.com/problems/long-pressed-name/
Author: Andrew Jarombek
Date: 5/23/2022
"""


def is_long_pressed_name(name: str, typed: str) -> bool:
    i = 0
    j = 0
    prev = None

    while i < len(name) and j < len(typed):
        c = name[i]

        if c != prev:
            while j < len(typed) and typed[j] == prev:
                j += 1

        if j == len(typed) or c != typed[j]:
            return False

        prev = c
        i += 1
        j += 1

    while j < len(typed) and typed[j] == prev:
        j += 1

    return j == len(typed) and i == len(name)
